From a791c287c37408938755ccb6b8b64f2ac04a5736 Mon Sep 17 00:00:00 2001
From: alpheratz0 <alpheratz99@protonmail.com>
Date: Thu, 9 Feb 2023 13:48:53 -0300
Subject: [PATCH] interpolate points when drawing

---
 sinca.c | 40 +++++++++++++++++++++++++++++++++++++++-
 1 file changed, 39 insertions(+), 1 deletion(-)

diff --git a/sinca.c b/sinca.c
index 406644c..13dafba 100644
--- a/sinca.c
+++ b/sinca.c
@@ -31,6 +31,8 @@
 #include "utils.h"
 #include "pizarra.h"
 
+#define swap(type, a, b) do { type ____a = a; a = b; b = ____a; } while (0)
+
 typedef struct {
 	bool active;
 	int x;
@@ -41,6 +43,7 @@ typedef struct {
 	bool active;
 	uint32_t color;
 	int brush_size;
+	xcb_point_t last_point;
 } DrawInfo;
 
 static Pizarra *pizarra;
@@ -250,6 +253,36 @@ addpoint(int x, int y, uint32_t color, int size)
 	}
 }
 
+static void
+addline(int x1, int y1, int x2, int y2, uint32_t color, int size)
+{
+	int x, y, dx, dy;
+
+	if (x1 == x2 && y1 == y2)
+		return;
+
+	dx = x2 - x1;
+	dy = y2 - y1;
+
+	if (abs(dx) > abs(dy)) {
+		if (x1 > x2) {
+			swap(int, x1, x2);
+			swap(int, y1, y2);
+		}
+
+		for (x = x1; x <= x2; x += size)
+			addpoint(x, dy * (x - x1) / dx + y1, color, size);
+	} else {
+		if (y1 > y2) {
+			swap(int, x1, x2);
+			swap(int, y1, y2);
+		}
+
+		for (y = y1; y <= y2; y += size)
+			addpoint(dx * (y - y1) / dy + x1, y, color, size);
+	}
+}
+
 static void
 h_button_press(xcb_button_press_event_t *ev)
 {
@@ -258,6 +291,8 @@ h_button_press(xcb_button_press_event_t *ev)
 		if (draginfo.active)
 			break;
 		drawinfo.active = true;
+		drawinfo.last_point.x = ev->event_x;
+		drawinfo.last_point.y = ev->event_y;
 		addpoint(ev->event_x, ev->event_y, drawinfo.color, drawinfo.brush_size);
 		pizarra_render(pizarra);
 		break;
@@ -298,7 +333,10 @@ h_motion_notify(xcb_motion_notify_event_t *ev)
 	}
 
 	if (drawinfo.active) {
-		addpoint(ev->event_x, ev->event_y, drawinfo.color, drawinfo.brush_size);
+		addline(drawinfo.last_point.x, drawinfo.last_point.y,
+				ev->event_x, ev->event_y, drawinfo.color, drawinfo.brush_size);
+		drawinfo.last_point.x = ev->event_x;
+		drawinfo.last_point.y = ev->event_y;
 		pizarra_render(pizarra);
 	}
 }
-- 
2.39.1

